pdns/dnsdistdist/dnsdist-opentelemetry.hh

```diff
@@ -0,0 +1,318 @@
+ * produced as the result of such linking.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+#pragma once
+
+#include <string>
+#include <vector>
+
+#ifndef DISABLE_PROTOBUF
+#include "protozero-trace.hh"
+using TraceID = pdns::trace::TraceID;
+using SpanID = pdns::trace::SpanID;
+using AnyValue = pdns::trace::AnyValue;
+using TracesData = pdns::trace::TracesData;
+#else
+// Define the minimal things needed
+#include <variant>
+using TraceID = int;
+using SpanID = int;
+using AnyValue = std::variant<std::string, int>;
+using TracesData = int;
+#endif
+
+/*
+ * This namespace contains all the bits and pieces required to do OpenTelemetry
+ * traces in dnsdist. It is contained in this header and cc-file to ensure the rest
+ * of the code is not littered with #ifdefs for DISABLE_PROTOBUF. All functions and
+ * other public members can be safely called/manipulated in a non-protobuf build of
+ * dnsdist.
+ *
+ * The idea is inspired by the rec-eventrace.{cc,hh} files.
+ *
+ * Although the namespace contains dnsdist, it might be general enough to be
+ * reused (after renaming the namespace) by auth and recursor
+ */
+namespace pdns::trace::dnsdist
+{
+
+/**
+ * @class Tracer
+ * @brief This class holds a single trace instance
+ *
+ */
+class Tracer
+{
+public:
+  Tracer()
+  {
+#ifndef DISABLE_PROTOBUF
+    pdns::trace::random(d_traceid);
+#endif
+  }
+  ~Tracer() = default;
+  Tracer(const Tracer&) = delete;
+  Tracer& operator=(const Tracer) = delete;
+  Tracer& operator=(Tracer&&) = delete;
+  Tracer(Tracer&&) = delete;
+
+  /**
+   * @brief Activate the Tracer
+   *
+   * Once activated, all new events are stored as actual Spans
+   */
+  void activate()
+  {
+#ifndef DISABLE_PROTOBUF
+    d_activated = true;
+    if (d_postActivationSpans == nullptr) {
+      d_postActivationSpans = std::make_unique<std::vector<pdns::trace::Span>>();
+    }
+#endif
+  }
+
+  /**
+   * @brief Deactivate the tracer
+   */
+  void deactivate()
+  {
+#ifndef DISABLE_PROTOBUF
+    // Set deactivated, but don't delete any data.
+    // The Tracer can be activated later
+    d_activated = false;
+#endif
+  }
+
+  /**
+   * @brief Add an attribute to the Trace
+   *
+   * @param key
+   * @param value
+   */
+  void setTraceAttribute(const std::string& key, const AnyValue& value);
+
+  /**
+   * @brief Create a new Span
+   *
+   * The Span's start time is set to the current time
+   *
+   * @param name The name for this span
+   * @return The SpanID of the created Span
+   */
+  SpanID addSpan(const std::string& name);
```

Comment:
It's not clear to me what is the use-case for calling this rather than `openSpan()`? I'm wondering if `addSpan` should be private. If the caller needs the ID it is possible to get it from the `Closer` object.

Reply:
I realize that `closeSpan` can't be private, `Closer` uses it.